ipld/graphsync/src/response_manager/link_tracker.rs

```diff
@@ -0,0 +1,292 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::RequestID;
+use cid::Cid;
+use std::collections::{HashMap, HashSet};
+
+/// LinkTracker records links being traversed to determine useful information
+/// in crafting responses for a peer. Specifically, if any in progress request
+/// has already sent a block for a given link, don't send it again.
+/// Second, keep track of whether links are missing blocks so you can determine
+/// at the end if a complete response has been transmitted.
+#[derive(Default)]
+pub struct LinkTracker {
+    /// The links traversed by any given request that were present.
+    present_blocks: HashMap<RequestID, Vec<Cid>>,
```

Comment:
I think you're right that this is only for child iplds! "traverse" doesn't really imply that the lookup was successful though, right? Otherwise `record_link_traversal` should probably be renamed as well